unison-cli/src/Unison/Cli/Monad.hs

```diff
@@ -0,0 +1,349 @@
+          loopState {currentPathStack = path List.NonEmpty.:| paths}
+      )
+
+-- | Create an initial loop state given a root branch and the current path.
+loopState0 :: Branch IO -> Path.Absolute -> LoopState
+loopState0 b p =
+  LoopState
+    { root = b,
+      lastSavedRoot = b,
+      currentPathStack = pure p,
+      latestFile = Nothing,
+      latestTypecheckedFile = Nothing,
+      lastInput = Nothing,
+      numberedArgs = []
+    }
+
+-- | Run a @Cli@ action down to @IO@.
+runCli :: Env -> Cli a a -> IO (ReturnType a)
+runCli env (Cli action) =
+  action env (\x -> pure (Success x))
+
+-- | Get the loop state.
+getLoopState :: Cli r LoopState
+getLoopState = do
+  Env {loopStateRef} <- ask
+  liftIO (readIORef loopStateRef)
+
+-- | Put the loop state.
+putLoopState :: LoopState -> Cli r ()
+putLoopState newSt = do
+  Env {loopStateRef} <- ask
+  liftIO (writeIORef loopStateRef newSt)
+
+-- | Modify the loop state (not thread-safe).
+modifyLoopState :: (LoopState -> LoopState) -> Cli r ()
+modifyLoopState f = do
+  Env {loopStateRef} <- ask
+  liftIO (modifyIORef' loopStateRef f)
+
+-- | The result of calling 'loadSource'.
+data LoadSourceResult
+  = InvalidSourceNameError
+  | LoadError
+  | LoadSuccess Text
+
+-- | Lift an action of type @IO (Either e a)@, given a continuation for @e@.
+ioE :: IO (Either e a) -> (e -> Cli r a) -> Cli r a
+ioE action errK =
+  Cli \env k ->
+    action >>= \case
+      Left err -> unCli (errK err) env k
+      Right value -> k value
+
+short :: ReturnType r -> Cli r a
+short r = Cli \_k _env -> pure r
+
+-- | Short-circuit with a value. Returns @r@ to the nearest enclosing
+-- 'newBlock' (or 'runCli' if there is no enclosing 'newBlock').
+returnWith :: r -> Cli r a
+returnWith = short . Success
+
+-- | Short-circuit the processing of the current input.
+returnEarly :: Output -> Cli r a
+returnEarly x = do
+  respond x
+  returnEarlyWithoutOutput
+
+-- | Variant of 'returnEarly' that doesn't take a final output message.
+returnEarlyWithoutOutput :: Cli r a
+returnEarlyWithoutOutput =
+  short Continue
+
+-- | Stop processing inputs from the user.
+haltRepl :: Cli r a
+haltRepl = short HaltRepl
+
+-- | Wrap a continuation with 'Cli'.
+--
+-- Useful for resource acquisition:
+--
+-- @
+-- resource <- with (bracket acquire close)
+-- @
+--
+-- Delimit the scope of acquired resources with 'newBlock'.
+with :: (forall x. (a -> IO x) -> IO x) -> Cli r a
+with resourceK = Cli \_ -> resourceK
+
+-- | A variant of 'with' for the variant of bracketing function that may return a Left rather than call the provided
+-- continuation.
+withE :: (forall x. (a -> IO x) -> IO (Either e x)) -> (e -> Cli r a) -> Cli r a
+withE resourceK errK =
+  Cli \env k ->
+    resourceK k >>= \case
+      Left err -> unCli (errK err) env k
+      Right val -> pure val
+
+-- | Run the given action in a new block, which delimits the scope of any 'succeedWith', 'with'/'withE', and 'time'
```

Comment:
I think the fact that `newBlock` is the delimiter for all of `with`, `succeedWith` AND `time` might be a bit of a foot-gun;

E.g. what if I have a block like this:

```
do
  scratchFile <- with (withFile "scratch.u")
  doSomethingTimeIntensive
  useFile scratchFile
```

Then I want to time a block, I could see myself making this mistake (probably not in such a simple case, but in a real-world code situation it'd be easier to miss):

```
do
  newBlock do
    scratchFile <- with (withFile "scratch.u")
    doSomethingTimeIntensive
  useFile scratchFile
```

Now we free the file before using it :'( ;

You could make the same mistake using nested do-blocks too by returning the file outside of its scope, but it's considerably less likely IMO.